apollo-federation/src/merger/merge_input.rs

```diff
@@ -0,0 +1,605 @@
+use std::collections::HashSet;
+
+use apollo_compiler::Node;
+use apollo_compiler::ast::InputValueDefinition;
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::schema::Component;
+use apollo_compiler::schema::InputObjectType;
+
+use crate::error::CompositionError;
+use crate::error::FederationError;
+use crate::error::SubgraphLocation;
+use crate::merger::hints::HintCode;
+use crate::merger::merge::Merger;
+use crate::merger::merge::Sources;
+use crate::schema::position::InputObjectTypeDefinitionPosition;
+use crate::supergraph::CompositionHint;
+
+impl Merger {
+    #[allow(dead_code)]
+    pub(crate) fn merge_input(
+        &mut self,
+        sources: &Sources<Node<InputObjectType>>,
+        dest: &InputObjectTypeDefinitionPosition,
+    ) -> Result<(), FederationError> {
+        let added = self.add_fields_shallow(sources, dest);
+
+        for (dest_field, subgraph_fields) in added {
+            if let Some(field_def) = subgraph_fields.values().find_map(|f| f.as_ref()) {
+                if dest_field.try_get(self.merged.schema()).is_none() {
+                    dest_field.insert(&mut self.merged, field_def.clone())?;
+                }
+            }
+
+            let dest_field_def = dest_field.try_get(self.merged.schema());
+            let is_inaccessible = if let Some(dest_field_def) = dest_field_def {
+                if let Some(inaccessible_directive) =
+                    &self.inaccessible_directive_name_in_supergraph
+                {
+                    dest_field_def.directives.has(inaccessible_directive)
+                } else {
+                    false
+                }
+            } else {
+                false
+            };
```

Comment:
I think we can simplify it to something like this

```suggestion
            let is_inaccessible = dest_field_def.is_some_and(|dest_field_def| {
                matches!(&self.inaccessible_directive_name_in_supergraph, Some(inaccessible_directive) if dest_field_def.directives.has(&inaccessible_directive))
            });
```